### PR TITLE
Add support of NNCF Accuracy Aware training 

### PR DESCRIPTION
**Pull request by kshpv**

### CODE REVIEW

**Comment by kshpv** [comment]
based on https://github.com/openvinotoolkit/nncf/pull/796

File: torchreid/engine/builder.py [hunk 1]
```diff
@@ -9,80 +10,135 @@ def build_engine(cfg, datamanager, model, optimizer, scheduler,
...
-            label_smooth=cfg.loss.softmax.label_smooth,
-            conf_penalty=cfg.loss.softmax.conf_penalty
-        )
+    is_accuracy_aware_training = False
+        if nncf_metainfo is not None:
+            if nncf_metainfo.nncf_config.get('compression').get('accuracy_aware_training', None) is not None:
+                is_accuracy_aware_training = True
+
+                                      'softmax and am_softmax losses for data.type = image')
+    if cfg.data.type == 'image':
```
Comment (kprokofi) [inline torchreid/engine/builder.py]: we don't support other data types now and removed it from the default config

File: configs/classification/efficientnet_b0_int8_accuracy_aware.yml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  root: './'
+  sources: ['CIFAR100_train']
+  targets: ['CIFAR100_val']
+  height: 224
+  width: 224
+  norm_mean: [0.485, 0.456, 0.406]
+  norm_std: [0.229, 0.224, 0.225]
+  save_dir: 'output/efficient_b0/log'
+  workers: 6
+  transforms:
```
Comment (kprokofi) [inline configs/classification/efficientnet_b0_int8_accuracy_aware.yml:37]: we decide to use only random_flip + augmix (it already includes random_rotate)

File: torchreid/engine/builder.py (hunk 1 above)
Comment (kshpv) [reply torchreid/engine/builder.py]: Done

File: configs/classification/efficientnet_b0_int8_accuracy_aware.yml (hunk 2 above)
Comment (kshpv) [reply configs/classification/efficientnet_b0_int8_accuracy_aware.yml:37]: But it is original efficienet config just with nncf section

File: torchreid/engine/builder.py [hunk 3]
```diff
@@ -1,63 +1,115 @@
...
+                should_freeze_aux_models=should_freeze_aux_models,
+                nncf_metainfo=nncf_metainfo,
+                initial_lr=initial_lr,
+                use_ema_decay=cfg.train.ema.enable,
+                ema_decay=cfg.train.ema.ema_decay
+            )
+        else:
+            target_metric_name = 'top1'
+            engine = AccuracyAwareImageAMSoftmaxEngine(
+                datamanager,
```
Comment (kprokofi) [inline torchreid/engine/builder.py]: There are a lot of duplicate parameters here and in the original AMSoftmax and if we change something we should change it twice. Maybe combine these parameters in some way?

File: torchreid/integration/nncf/compression.py [hunk 4]
```diff
@@ -74,131 +52,94 @@ def _load_checkpoint_for_nncf(model, filename, map_location=None, strict=False):
...
+            targets = list(test_loader.keys())
+            use_gpu = True if 'cuda' == cur_device.type else False
+            for dataset_name in targets:
+                domain = 'source' if dataset_name in datamanager_for_init.sources else 'target'
+                print('##### Evaluating {} ({}) #####'.format(dataset_name, domain))
+                cmc, m_ap, norm_cm = evaluate_classification(test_loader[dataset_name]['query'], model,
+                                                             use_gpu=use_gpu, topk=(1, 5))
+
+            top1, top5 = cmc[0], cmc[1]
+            if target_metric == 'top1':
```
Comment (kprokofi) [inline torchreid/integration/nncf/compression.py:118]: You assigned the value above and check it here. Can the variable change somehow?

File: torchreid/integration/nncf/engine.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            if len(self.models) > 1:
+                return True
+            elif len(self.models) == 1:
+                return False
+            else:
+                raise RuntimeError('Incorrect number of models')
+
+        if not is_mutual_learning_enable():
+            model = self.models['model']
+            optimizer = self.optims['model']
```
Comment (kprokofi) [inline torchreid/integration/nncf/engine.py]: It's better to use attribute 'self.main_model_name' or method 'get_model_names'

File: torchreid/integration/nncf/engine.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            scheduler = self.scheds['model']
+        else:
+            # Use only main model
+            model = self.models['model_0']
+            optimizer = self.optims['model_0']
+            scheduler = self.scheds['model_0']
+
+        def _configure_optimizers_fn():
+            return optimizer, scheduler
```
Comment (kprokofi) [inline torchreid/integration/nncf/engine.py:18]: How could we get around this behavior using non-local variables in this scope?

**Review by kprokofi — COMMENTED** [review]
First comments

File: torchreid/engine/builder.py (hunk 3 above)
Comment (sovrasov) [reply torchreid/engine/builder.py]: This could be done by creating a dict with params in advance and passing it as kwargs in both the branches.

File: torchreid/integration/nncf/engine.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+from torchreid.engine.image import ImageAMSoftmaxEngine
+
+
+class AccuracyAwareImageAMSoftmaxEngine(ImageAMSoftmaxEngine):
+    def __init__(self, datamanager, models, optimizers, reg_cfg, metric_cfg, schedulers=None, use_gpu=False, save_chkpt=True,
```
Comment (sovrasov) [inline torchreid/integration/nncf/engine.py]: `nncf_arg1, nncf_arg1, **kwargs` then `super().__init__(**kwargs)` ?

File: torchreid/engine/builder.py (hunk 3 above)
Comment (kshpv) [reply torchreid/engine/builder.py]: Done

File: torchreid/integration/nncf/engine.py (hunk 5 above)
Comment (kshpv) [reply torchreid/integration/nncf/engine.py]: Done

File: torchreid/integration/nncf/compression.py (hunk 4 above)
Comment (kshpv) [reply torchreid/integration/nncf/compression.py:118]: Probably, yes

File: torchreid/integration/nncf/engine.py (hunk 7 above)
Comment (kshpv) [reply torchreid/integration/nncf/engine.py]: Done

**Comment by kprokofi** [comment]
@lzrvch @LeonidBeynenson , Do you have some comments or we can merge?

**Comment by kshpv** [comment]
> @lzrvch @LeonidBeynenson , Do you have some comments or we can merge?

Please don't merge till https://github.com/openvinotoolkit/nncf/pull/796 is not merged.

**Comment by kshpv** [comment]
The PR is ready for merge. Please, merge if satisfied

**Comment by daniil-lyakhov** [comment]
Can't launch mobilenet_v3 training due to this line https://github.com/openvinotoolkit/deep-object-reid/blob/ak/accuracy_aware/torchreid/models/mobilenetv3.py#L142

File: torchreid/engine/image/am_softmax.py [hunk 8]
```diff
@@ -264,8 +264,12 @@ def forward_backward(self, data):
 
 
             total_loss.backward(retain_graph=self.enable_metric_losses)
-
             for model_name in model_names:
+                # TODO(kshpv, lbeynens): Change this ugly code. The bug appears trying to
+                # call optimzer with mutual learning and NNCF
+                if model_name != 'model_0':
+                    if self._should_freeze_aux_models(self.epoch):
+                        continue
```
Comment (daniil-lyakhov) [inline torchreid/engine/image/am_softmax.py:272]: This won't work because model_names were changed

Comment (daniil-lyakhov) [reply torchreid/engine/image/am_softmax.py:272]:
```suggestion
                # TODO(kshpv, lbeynens): Change this ugly code. The bug appears trying to
                # call optimzer with mutual learning and NNCF
                if not self.models[model_name].training:
                    continue
```

File: torchreid/models/mobilenetv3.py [hunk 9]
```diff
@@ -25,10 +24,6 @@
 }
 
 
-SHOULD_NNCF_SKIP_SE_LAYERS = False
-SHOULD_NNCF_SKIP_HEAD = False
-no_nncf_se_layer_context = get_no_nncf_trace_context_manager() if SHOULD_NNCF_SKIP_SE_LAYERS else nullcontext
-no_nncf_head_context = get_no_nncf_trace_context_manager() if SHOULD_NNCF_SKIP_HEAD else nullcontext
```
Comment (daniil-lyakhov) [inline torchreid/models/mobilenetv3.py:31]: Looks like you didn't remove all references to `no_nncf_head_context`

**Comment by daniil-lyakhov** [comment]
Consider using this fix https://github.com/daniil-lyakhov/deep-object-reid/blob/dl/update_nncf/torchreid/utils/tools.py#L155-L160
